### PR TITLE
[build] Install extra utility packages

### PR DESCRIPTION
**Pull request by saimidu**
*Issue #, if available:*

## PR Checklist
- [x] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [x] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [x] (If applicable) I've documented below the tests I've run on the DLC image
- [x] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [x] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Reviewer Checklist
* For reviewer, before merging, please cross-check:
- [ ] I've verified the code change on the config file mentioned above has already been reverted

*Description:*
Install extra utility packages on training DLC images

*Tests run:*

*DLC image/dockerfile:*

*Additional context:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/dlc_tests/sanity/test_utility_installation.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+    """
+    ctx = Context()
+    container_name = test_utils.get_container_name("plotly", training)
+    test_utils.start_container(container_name, training, ctx)
+
+    framework = test_utils.get_framework_from_image_uri(training)
+
+    package_list_cmd = "conda list" if "pytorch" in framework else "pip freeze"
+
+    test_utils.run_cmd_on_container(container_name, ctx, f"{package_list_cmd} | grep -i plotly")
```
Comment (jeet4320) [inline test/dlc_tests/sanity/test_utility_installation.py:168]: can we club together all packages import in single function (probably make a list of packages and loop through to check their imports), its seem repetitive to have func for every package with similar code.

File: tensorflow/training/docker/1.15/py3/cu110/Dockerfile.gpu [hunk 2]
```diff
@@ -166,7 +166,7 @@ RUN ${PIP} install --no-cache-dir -U \
     scipy==1.5.2 \
     scikit-learn==0.23 \
     pandas==1.1 \
-    Pillow==7.2.0 \
+    Pillow==8.2.0 \
     python-dateutil==2.8.1 \
     # install PyYAML>=5.4,<5.5 to avoid conflict with latest awscli
     "pyYAML>=5.4,<5.5" \
```
Comment (jeet4320) [inline tensorflow/training/docker/1.15/py3/cu110/Dockerfile.gpu:172]: this is missing libgtk2.0-dev

File: tensorflow/training/docker/1.15/py3/Dockerfile.cpu [hunk 3]
```diff
@@ -140,6 +142,15 @@ RUN ${PIP} install --no-cache-dir -U \
     "pyYAML>=5.4,<5.5" \
     "awscli<2"
 
+# Install extra packages
+RUN pip install --no-cache-dir -U \
+    bokeh \
+    imageio \
+    opencv-python==4.3.0.36 \
```
Comment (jeet4320) [inline tensorflow/training/docker/1.15/py3/Dockerfile.cpu:149]: why its pinned

**Comment by jeet4320** [comment]
TF, MXnet and PT have passed on different commits.

File: mxnet/training/docker/1.8.0/py3/Dockerfile.cpu [hunk 4]
```diff
@@ -131,6 +133,15 @@ RUN ${PIP} install --no-cache --upgrade \
...
     smclarify
 
+# Install extra packages
+RUN pip install --no-cache-dir -U \
+    bokeh \
+    imageio \
+    opencv-python \
+    plotly \
+    seaborn \
+    shap
```
Comment (saimidu) [inline mxnet/training/docker/1.8.0/py3/Dockerfile.cpu:143]: Could you please constrain the packages to the following versions on all dockerfiles:
```
"bokeh>=2.3,<3"
"imageio>=2.9,<3"
"opencv-python>=4.3,<5"
"plotly>=5.1,<6"
"seaborn>=0.11,<1"
"shap>=0.39,<1"
```

File: tensorflow/training/docker/1.15/py3/Dockerfile.cpu (hunk 3 above)
Comment (saimidu) [inline tensorflow/training/docker/1.15/py3/Dockerfile.cpu:149]: Can we unpin this from 4.3.0 and allow it to be anywhere between `>=4.0,<5`?

File: tensorflow/training/docker/1.15/py36/Dockerfile.cpu [hunk 5]
```diff
@@ -146,6 +148,15 @@ RUN pip install --no-cache-dir -U \
     smdebug==${SMDEBUG_VERSION} \
     smclarify
 
+# Install extra packages
+RUN pip install --no-cache-dir -U \
+    bokeh \
+    imageio \
+    opencv-python==4.3.0.36 \
```
Comment (saimidu) [inline tensorflow/training/docker/1.15/py36/Dockerfile.cpu:155]: Can we unpin this from 4.3.0 and allow it to be anywhere between `>=4.0,<5`?

File: tensorflow/training/docker/1.15/py36/cu100/Dockerfile.gpu [hunk 6]
```diff
@@ -185,6 +187,14 @@ RUN pip install --no-cache-dir -U \
     "awscli<2" \
     smdebug==${SMDEBUG_VERSION}
 
+# Install extra packages
+RUN pip install --no-cache-dir -U \
+    bokeh \
+    imageio \
+    opencv-python==4.3.0.36 \
```
Comment (saimidu) [inline tensorflow/training/docker/1.15/py36/cu100/Dockerfile.gpu:194]: Can we unpin this from 4.3.0 and allow it to be anywhere between `>=4.0,<5`?

File: test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py [hunk 7]
```diff
@@ -179,19 +179,19 @@ def test_pytorch_amp(pytorch_training, ec2_connection, gpu_only, ec2_instance_ty
...
 @pytest.mark.model("N/A")
 @pytest.mark.parametrize("ec2_instance_type", PT_EC2_CPU_INSTANCE_TYPE, indirect=True)
-def test_pytorch_s3_plugin_cpu(pytorch_training, ec2_connection, cpu_only, ec2_instance_type, pt16_and_above_only):
-    if test_utils.is_image_incompatible_with_instance_type(pytorch_training, ec2_instance_type):
-        pytest.skip(f"Image {pytorch_training} is incompatible with instance type {ec2_instance_type}")
-    execute_ec2_training_test(ec2_connection, pytorch_training, PT_S3_PLUGIN_CMD)
+def test_pytorch_s3_plugin_cpu(pytorch_training, ec2_connection, cpu_only, ec2_instance_type):
+    _, image_framework_version = get_framework_and_version_from_tag(pytorch_training)
+    if Version(image_framework_version) < Version("1.8"):
+        pytest.skip("S3 plugin is supported on PyTorch version >=1.8")
```
Comment (saimidu) [inline test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py:194]: Please undo this change.

File: test/dlc_tests/ecs/pytorch/training/test_ecs_pytorch_training.py [hunk 8]
```diff
@@ -71,6 +71,9 @@ def test_ecs_pytorch_s3_plugin_training_cpu(cpu_only, ecs_container_instance, py
     Given above parameters, registers a task with family named after this test, runs the task, and waits for
     the task to be stopped before doing teardown operations of instance and cluster.
     """
+    _, image_framework_version = get_framework_and_version_from_tag(pytorch_training)
+    if Version(image_framework_version) < Version("1.8"):
+        pytest.skip("S3 plugin is supported on PyTorch version >=1.8")
```
Comment (saimidu) [inline test/dlc_tests/ecs/pytorch/training/test_ecs_pytorch_training.py:76]: Please undo this, and allow this to be fixed in the PR dedicated to this.

File: test/dlc_tests/ecs/pytorch/training/test_ecs_pytorch_training.py [hunk 9]
```diff
@@ -91,6 +94,9 @@ def test_ecs_pytorch_s3_plugin_training_gpu(gpu_only, ecs_container_instance, py
...
```
Comment (saimidu) [inline test/dlc_tests/ecs/pytorch/training/test_ecs_pytorch_training.py:99]: Please undo this, and allow this to be fixed in the PR dedicated to this.

File: test/dlc_tests/eks/pytorch/training/test_eks_pytorch_training.py [hunk 10]
```diff
@@ -103,6 +103,9 @@ def test_eks_pt_s3_plugin_single_node_training(pytorch_training):
         :param setup_utils: environment in which EKS tools are setup
         :param pytorch_training: the ECR URI
     """
+    _, image_framework_version = get_framework_and_version_from_tag(pytorch_training)
+    if Version(image_framework_version) < Version("1.8"):
+        pytest.skip("S3 plugin is supported on PyTorch version >=1.8")
```
Comment (saimidu) [inline test/dlc_tests/eks/pytorch/training/test_eks_pytorch_training.py:108]: Please undo this, and allow this to be fixed in the PR dedicated to this.

File: test/dlc_tests/sanity/test_pt_s3.py [hunk 11]
```diff
@@ -14,12 +14,14 @@
...
 @pytest.mark.model("N/A")
-def test_pt_s3_sanity(pytorch_training, pt16_and_above_only):
+def test_pt_s3_sanity(pytorch_training):
     """
     Check that the internally built PT S3 binary is properly installed.
     :param pytorch_training: framework fixture for pytorch training
     """
     _, framework_version = get_framework_and_version_from_tag(pytorch_training)
+    if Version(framework_version) < Version("1.8"):
+        pytest.skip("S3 plugin sanity is supported on PyTorch version >=1.8")
```
Comment (saimidu) [inline test/dlc_tests/sanity/test_pt_s3.py:24]: Please undo this, and allow this to be fixed in the PR dedicated to this.

File: test/dlc_tests/sanity/test_utility_installation.py [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+    :param training: training ECR image URI
+    """
+    ctx = Context()
+    container_name = test_utils.get_container_name("utility_packages_using_import", training)
+    test_utils.start_container(container_name, training, ctx)
+
+    framework, image_framework_version = test_utils.get_framework_and_version_from_tag(training)
+    if "mxnet" in framework:
+        if Version(image_framework_version) <= Version("1.8.0"):
+            pytest.skip("Extra utility packages will be added going forward")
```
Comment (saimidu) [inline test/dlc_tests/sanity/test_utility_installation.py:53]: Please make this a dictionary for reference, rather than 3 different if conditions, for readability.

Example:

```
utility_package_minimum_framework_version = {
    "mxnet": "1.8",
    "pytorch": "1.7",
    "tensorflow2": "2.4",
    "tensorflow1": "1.15",
}
framework, framework_version = get_framework_and_version_from_tag(image)
framework = "tensorflow1" if framework == "tensorflow" and framework_version.startswith("1.") else "tensorflow2"
if Version(framework_version) < Version(utility_package_minimum_framework_version[framework]):
    pytest.skip("Extra utility packages will be added going forward.")
```

File: test/dlc_tests/sanity/test_utility_installation.py [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+
+    if "tensorflow" in framework:
+        if Version(image_framework_version) <= Version("2.4.0") or Version(image_framework_version) == Version("1.15"):
```
Comment (saimidu) [inline test/dlc_tests/sanity/test_utility_installation.py:56]: We want to test TF 1.15 images. This condition will skip TF 1.15 DLC images for this test.

File: test/dlc_tests/sanity/test_utility_installation.py [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+            pytest.skip("Extra utility packages will be added going forward")
+
+    if "pytorch" in framework:
+        if Version(image_framework_version) <= Version("1.7.1"):
+            pytest.skip("Extra utility packages will be added going forward")
+
+    package_list_cmd = "conda list" if "pytorch" in framework else "pip freeze"
+    
+    for package in UTILITY_PACKAGES_IMPORT:
+        if Version(image_framework_version) == Version("1.5.0") and package == "sagemaker":
```
Comment (saimidu) [inline test/dlc_tests/sanity/test_utility_installation.py:66]: We should also check here that `framework == "pytorch"`.

File: test/dlc_tests/sanity/test_utility_installation.py [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+
+    if "tensorflow" in framework:
+        if Version(image_framework_version) <= Version("2.4.0") or Version(image_framework_version) == Version("1.15"):
+            pytest.skip("Extra utility packages will be added going forward")
+
+    if "pytorch" in framework:
+        if Version(image_framework_version) <= Version("1.7.1"):
+            pytest.skip("Extra utility packages will be added going forward")
+
+    package_list_cmd = "conda list" if "pytorch" in framework else "pip freeze"
```
Comment (saimidu) [inline test/dlc_tests/sanity/test_utility_installation.py:63]: If `package_list_cmd` is unused, then this line can be deleted.

File: test/dlc_tests/sanity/test_pt_s3.py (hunk 11 above)
Comment (jeet4320) [reply test/dlc_tests/sanity/test_pt_s3.py:24]: I have added this as tests fail if its not there. Other PR can remove it after merging mainline.
Without this code, it blocks the PT1.7.1 release if there is some issue

File: tensorflow/training/docker/1.15/py3/cu110/Dockerfile.gpu [hunk 16]
```diff
@@ -173,7 +173,7 @@ RUN ${PIP} install --no-cache-dir -U \
     requests==2.24.0 \
     "awscli<2" \
     mpi4py==3.0.3 \
-    opencv-python==4.3.0.36 \
+    opencv-python>=4.0,<5 \
```
Comment (saimidu) [inline tensorflow/training/docker/1.15/py3/cu110/Dockerfile.gpu:176]: Please change this to `"opencv-python>=4.3,<5" \` to resolve build issues.

File: mxnet/training/docker/1.8.0/py3/Dockerfile.cpu [hunk 17]
```diff
@@ -135,12 +135,12 @@ RUN ${PIP} install --no-cache --upgrade \
...
-    opencv-python \
-    plotly \
-    seaborn \
-    shap
+   "bokeh>=2.3,<3" \
+   "imageio>=2.9,<3" \
+   "opencv-python>=4.3,<5" \
+   "plotly>=5.1,<6" \
+   "seaborn>=0.11,<1" \
+   "shap>=0.39,<1"
```
Comment (saimidu) [inline mxnet/training/docker/1.8.0/py3/Dockerfile.cpu:143]: nit: Please fix indentation by placing 4 spaces before the start of `"..." \` in each line here.

File: tensorflow/training/docker/1.15/py3/cu110/Dockerfile.gpu [hunk 18]
```diff
@@ -189,12 +189,12 @@ RUN ${PIP} install --no-cache-dir -U \
...
 RUN pip install --no-cache-dir -U \
-    bokeh \
-    imageio \
-    plotly \
-    seaborn \
-    shap
-
+   "bokeh>=2.3,<3" \
+   "imageio>=2.9,<3" \
+   "opencv-python>=4.3,<5" \
```
Comment (saimidu) [inline tensorflow/training/docker/1.15/py3/cu110/Dockerfile.gpu:193]: Please delete `opencv-python` from here, as it is already installed in line 176.

File: test/dlc_tests/sanity/test_utility_installation.py [hunk 19]
```diff
@@ -48,23 +48,18 @@ def test_utility_packages_using_import(training):
...
+    framework, framework_version = get_framework_and_version_from_tag(image)
+    framework = "tensorflow1" if framework == "tensorflow" and framework_version.startswith("1.") else "tensorflow2"
+    if Version(framework_version) < Version(utility_package_minimum_framework_version[framework]):
+        pytest.skip("Extra utility packages will be added going forward.")
     
     for package in UTILITY_PACKAGES_IMPORT:
-        if Version(image_framework_version) == Version("1.5.0") and package == "sagemaker":
-            pytest.skip("sagemaker version < 2.0 is installed for PT 1.5.0 images")
         version = test_utils.run_cmd_on_container(container_name, ctx, f"import {package}; print({package}.__version__)", executable="python").stdout.strip()
         if package == "sagemaker":
```
Comment (saimidu) [inline test/dlc_tests/sanity/test_utility_installation.py:64]:
```suggestion
        if package == "sagemaker" and not (framework == "pytorch" and Version(framework_version) < Version("1.6")):
```

Comment (jeet4320) [reply test/dlc_tests/sanity/test_utility_installation.py:64]: we are checking sagemaker for pytorch as well and execution will come on this like if PT is >=1.7, so no need to check for version

File: tensorflow/training/docker/1.15/py3/cu110/Dockerfile.gpu (hunk 18 above)
Comment (jeet4320) [reply tensorflow/training/docker/1.15/py3/cu110/Dockerfile.gpu:193]: I deleted from 176

File: mxnet/training/docker/1.8.0/py3/Dockerfile.cpu (hunk 17 above)
Comment (jeet4320) [reply mxnet/training/docker/1.8.0/py3/Dockerfile.cpu:143]: will do, will wait for CI and then will add these changes.

**Comment by saimidu** [comment]
Approved.
